[WIP] Cancel confirmed appointment feature

davinotdavid opened the pull request:
<!--
* Filling out the template is required.
* All new code must have been tested to ensure against regressions
-->

## Description of the Change

<!-- We must be able to understand the design of your change from this description, so please walk us through the concepts. -->
[**Main work**]
- In the frontend, added a textarea for an optional reason and a `Cancel booking` button in the `AppointmentModal` component
- In the backend, added a new endpoint to cancel an appointment (`POST` to `/apmt/{id}/cancel`) that looks up all the slots associated with that appointment and:
   1. Sends a cancellation email with and optional reason (new email template added)
   2. If the `meeting_link_provider` is Zoom, calls the Zoom client to delete the meeting

[**Small refactor**]
- We were using `reject` and `cancel` a little bit interchangeably which creates some confusion (at least for me) in terms of _when_ that action took place. For this PR, I have separated the concepts so that `reject` means that an appointment that is pending approval has been rejected (as in, never accepted) and `cancelled` is cancelling an appointment that has already been confirmed.


[**Fix**]
- When a schedule is created and the settings for creating a zoom link is enabled, it gets set in the `meeting_link_provider` of the schedule itself. However, the appointment doesn't seem to ever get its `meeting_link_provider` updated so upon appointment creation, I've "synced" their fields so that I can later read it from appointment cancellation.



## Benefits

<!-- What benefits will be realized by the code change? -->

- New cancellation of confirmed booking feature that also deletes Zoom meetings if any
- Separation of cancel vs. reject concepts

## Applicable Issues

<!-- Enter any applicable issues here -->
https://github.com/thunderbird/appointment/issues/171
https://github.com/thunderbird/appointment/issues/1061

devmount commented on frontend/src/components/AppointmentModal.vue:
```diff
@@ -62,6 +64,13 @@ const deleteAppointment = () => {
+// Handle cancel
+const cancelAppointment = () => {
+  apmtStore.cancelAppointment(props.appointment?.id, cancelReason.value);
+  cancelReason.value = "";
```
See comment above

```suggestion
  cancelReason.value = '';
```

davinotdavid replied:
Thanks! Updated!

MelissaAutumn commented on backend/src/appointment/routes/api.py:
```diff
@@ -456,6 +458,45 @@ def delete_my_appointment(id: int, db: Session = Depends(get_db), subscriber: Su
 
 
+@router.post('/apmt/{id}/cancel')
+def cancel_my_appointment(
```
Like @devmount mentioned we'll need to call calendar.delete_event, we don't have a good one-line way to initialize that right now, so feel free to copy and paste from another function.

MelissaAutumn commented on backend/test/integration/test_appointment.py:
```diff
@@ -124,3 +129,92 @@ def test_dont_show_other_subscribers_appointments(
+            assert response.status_code == 200
+            assert mock_tools().send_cancel_vevent.call_count == len(appointment.slots)
+
+        db = with_db()
```
You can do `with with_db() as db:` to avoid having to call .close(). I do recognize that's maybe one with too many though 😅

MelissaAutumn commented on frontend/src/components/AppointmentModal.vue:
```diff
@@ -176,8 +185,23 @@ const deleteAppointment = () => {
+        <p class="mb-8">This booking is confirmed.</p>
+        <form class="gap-4" @submit.prevent>
+          <label for="cancelReason">
+            {{ t("label.cancelReason") }}
```
```suggestion
            {{ t('label.cancelReason') }}
```

davinotdavid commented on frontend/src/stores/appointment-store.ts:
For sure, yes! I've added one just to assert that the call is being done to the correct endpoint with the expected body. Is it similar with what you were thinking for this test?

devmount replied:
Awesome, thanks!

davinotdavid commented on frontend/src/components/AppointmentModal.vue:
```diff
@@ -38,6 +38,8 @@ interface Props {
 }
 const props = defineProps<Props>();
 
+const cancelReason = ref<string>("");
```
Oh thank you! I will keep that in mind for future PRs, good catch! 🙏

davinotdavid commented on backend/test/integration/test_appointment.py:
Thank you so much! I really appreciate your reviews 🙏

davinotdavid commented:
@MelissaAutumn @devmount Thank you so much for the review / insights! I really appreciate it 🙇

I've updated the PR description with the new commits but the diff / changes are:
- It is now getting the remote calendar connection and calling `delete_event`
- Added a check on the frontend for `isPast` to not render the cancellation form if the appointment is in the past

rwood-moz commented on backend/src/appointment/templates/email/cancelled.jinja2:
```diff
@@ -0,0 +1,9 @@
+{% block introduction %}
+  <div style="color: {{ colour_text_base }}; text-align: center; line-height: 16px;">
+    <p>
+      {{ l10n('cancel-mail-html-heading', {'owner_name': owner_name, 'date': date, 'reason_line': reason_line}) }}
```
Oh wait, nvm, that builds the heading with those params hah

davinotdavid replied:
Good point! It might be a bit unclear but the intention was to build the `reason_line` as `Reason: { $reason }` but I figured that `Reason` should also be in i18n so I resorted for calling it `reason_line` and pass it to the templates instead 😅

rwood-moz commented:
> Thanks for the quick implementation of the review feedback! Looks good on my end now! @MelissaAutumn @rwood-moz ?

LGTM! Just one question (that doesn't need to be addressed in this PR) - after an appointment is cancelled there is no history of it. Do we want cancelled bookings to appear on the `Booking` tab => `Past` appointment list, and show a status of cancelled? What if a user cancels the wrong appointment but they aren't able to confirm as it just disappears from the bookings history? Not sure if that should be considered.
